private/bufpkg/bufcheck/buflint/internal/buflintvalidate/numeric.go

```diff
@@ -0,0 +1,274 @@
+	)
+}
+
+func checkNumericRules[
+	T int32 | int64 | uint32 | uint64 | float32 | float64 | timestamppb.Timestamp | durationpb.Duration,
+](
+	adder *adder,
+	ruleFieldNumber int32,
+	ruleMessage protoreflect.Message,
+	// convertFunc returns the converted value, a file annotation string and an error.
+	convertFunc func(protoreflect.Value) (*T, string, error),
+	// compareFunc returns a positive value if the first argument is bigger,
+	// a negative value if the second argument is bigger or 0 if they are equal.
+	compareFunc func(*T, *T) float64,
+	formatFunc func(*T) interface{},
+) error {
+	var constant, lowerBound, gt, gte, upperBound, lt, lte *T
+	var in, notIn []*T
+	var fieldCount int
+	var constFieldNumber, inFieldNumber, lowerBoundFieldNumber, upperBoundFieldNumber int32
+	var err error
+	ruleMessage.Range(func(field protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+		fieldCount++
+		var convertErrorMessage string
+		fieldNumber := int32(field.Number())
+		switch fieldName := string(field.Name()); fieldName {
+		case "const":
+			constFieldNumber = fieldNumber
+			constant, convertErrorMessage, err = convertFunc(value)
+		case "gt":
+			gt, convertErrorMessage, err = convertFunc(value)
+			lowerBound = gt
+			lowerBoundFieldNumber = fieldNumber
+		case "gte":
+			gte, convertErrorMessage, err = convertFunc(value)
+			lowerBound = gte
+			lowerBoundFieldNumber = fieldNumber
+		case "lt":
+			lt, convertErrorMessage, err = convertFunc(value)
+			upperBound = lt
+			upperBoundFieldNumber = fieldNumber
+		case "lte":
+			lte, convertErrorMessage, err = convertFunc(value)
+			upperBound = lte
+			upperBoundFieldNumber = fieldNumber
+		case "in":
+			inFieldNumber = fieldNumber
+			for i := 0; i < value.List().Len(); i++ {
+				var converted *T
+				converted, convertErrorMessage, err = convertFunc(value.List().Get(i))
+				if converted != nil {
+					in = append(in, converted)
+				}
+			}
+		case "not_in":
+			for i := 0; i < value.List().Len(); i++ {
+				var converted *T
+				converted, convertErrorMessage, err = convertFunc(value.List().Get(i))
+				if converted != nil {
+					notIn = append(notIn, converted)
+				}
+			}
+		}
+		if convertErrorMessage != "" {
+			adder.addForPathf(
+				[]int32{ruleFieldNumber, fieldNumber},
+				"Field %q has an invalid %s: %s.",
+				adder.fieldName(),
+				adder.getFieldRuleName(ruleFieldNumber, int32(field.Number())),
+				convertErrorMessage,
+			)
+		}
+		return true
+	})
+	if err != nil {
+		return err
+	}
+	if constant != nil && fieldCount > 1 {
+		adder.addForPathf(
+			[]int32{ruleFieldNumber, constFieldNumber},
+			"Field %q has %s and does not need other rules in %s.",
+			adder.fieldName(),
+			adder.getFieldRuleName(ruleFieldNumber, constFieldNumber),
+			adder.getFieldRuleName(ruleFieldNumber),
+		)
+	}
+	if len(in) > 0 && fieldCount > 1 {
+		adder.addForPathf(
+			[]int32{ruleFieldNumber, inFieldNumber},
+			"Field %q has %s and does not need other rules in %s.",
```

Comment:
Can you give an example of an error message printed out here?

Reply:
`Field "fld" has (buf.validate.field).sint32.in and does not need other rules in (buf.validate.field).sint32.`

Reply:
`Field %q has %s, which overrides all other fields for %s. Remove other fields for %s.`

Reply:
Actually, `in` doesn't override other rules. The other rules can be simplified. I was thinking of something like `Field "fld" has (buf.validate.field).int32.in, therefore other rules in  (buf.validate.field).int32 should be simplified by removing allowed values from  (buf.validate.field).int32.in that fail these rules.` What do you think?

Reply:
Discussed offline on Friday, this check is no longer needed.